core/src/main/java/org/apache/iceberg/metrics/TimerResultParser.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.iceberg.util.JsonUtil;
+
+class TimerResultParser {
+  private static final String MISSING_FIELD_ERROR_MSG =
+      "Cannot parse timer from '%s': Missing field '%s'";
+
+  private static final String NAME = "name";
+  private static final String TIME_UNIT = "time-unit";
+  private static final String COUNT = "count";
+  private static final String TOTAL_DURATION = "total-duration";
+
+  private TimerResultParser() {}
+
+  static String toJson(TimerResult timerResult) {
+    return toJson(timerResult, false);
+  }
+
+  static String toJson(TimerResult timerResult, boolean pretty) {
+    return JsonUtil.generate(gen -> toJson(timerResult, gen), pretty);
+  }
+
+  static void toJson(TimerResult timer, JsonGenerator gen) throws IOException {
+    Preconditions.checkArgument(null != timer, "Invalid timer: null");
+
+    gen.writeStartObject();
+    gen.writeStringField(NAME, timer.name());
+    gen.writeNumberField(COUNT, timer.count());
+    gen.writeStringField(TIME_UNIT, timer.timeUnit().name());
+    gen.writeNumberField(TOTAL_DURATION, fromDuration(timer.totalDuration(), timer.timeUnit()));
+    gen.writeEndObject();
+  }
+
+  static TimerResult fromJson(String json) {
+    return JsonUtil.parse(json, TimerResultParser::fromJson);
+  }
+
+  static TimerResult fromJson(String property, JsonNode json) {
+    return fromJson(get(property, json));
+  }
+
+  static TimerResult fromJson(JsonNode json) {
+    Preconditions.checkArgument(null != json, "Cannot parse timer from null object");
+    Preconditions.checkArgument(json.isObject(), "Cannot parse timer from non-object: %s", json);
+
+    String name = JsonUtil.getString(NAME, json);
+    long count = JsonUtil.getLong(COUNT, json);
+    TimeUnit unit = TimeUnit.valueOf(JsonUtil.getString(TIME_UNIT, json));
+    long duration = JsonUtil.getLong(TOTAL_DURATION, json);
+    return new TimerResult(name, unit, toDuration(duration, unit), count);
+  }
+
+  private static JsonNode get(String property, JsonNode node) {
+    Preconditions.checkArgument(
+        node.has(property), "Cannot parse timer from missing field: %s", property);
+
+    JsonNode timer = node.get(property);
+    Preconditions.checkArgument(timer.has(NAME), MISSING_FIELD_ERROR_MSG, property, NAME);
+    Preconditions.checkArgument(timer.has(COUNT), MISSING_FIELD_ERROR_MSG, property, COUNT);
+    Preconditions.checkArgument(timer.has(TIME_UNIT), MISSING_FIELD_ERROR_MSG, property, TIME_UNIT);
+    Preconditions.checkArgument(
+        timer.has(TOTAL_DURATION), MISSING_FIELD_ERROR_MSG, property, TOTAL_DURATION);
+    return timer;
+  }
+
+  @VisibleForTesting
+  static long fromDuration(Duration duration, TimeUnit unit) {
+    return unit.convert(duration.toNanos(), TimeUnit.NANOSECONDS);
+  }
+
+  @VisibleForTesting
+  static Duration toDuration(long val, TimeUnit unit) {
+    return Duration.of(val, toChronoUnit(unit));
+  }
+
+  private static ChronoUnit toChronoUnit(TimeUnit unit) {
+    // copied from JDK9
```

Comment:
If this is copied from elsewhere, we need to note it in the LICENSE file and document what this is licensed under.

On the other hand, if you were to delete this and write it without reference to the original, I don't think that we would need to update the license documentation.

Reply:
ok removed the comment